adder/adder.go

```diff
@@ -2,23 +2,131 @@ package adder
 
 import (
 	"context"
+	"io"
 	"mime/multipart"
 
+	"github.com/ipfs/ipfs-cluster/adder/ipfsadd"
 	"github.com/ipfs/ipfs-cluster/api"
 
 	cid "github.com/ipfs/go-cid"
+	files "github.com/ipfs/go-ipfs-cmdkit/files"
+	ipld "github.com/ipfs/go-ipld-format"
 	logging "github.com/ipfs/go-log"
 )
 
 var logger = logging.Logger("adder")
 
-// Adder represents a module capable of adding content to IPFS Cluster.
-type Adder interface {
-	// FromMultipart adds from a multipart reader and returns
-	// the resulting CID.
-	FromMultipart(context.Context, *multipart.Reader, *api.AddParams) (*cid.Cid, error)
+// ClusterDAGService is an implementation of ipld.DAGService plus a Finalize
+// method. ClusterDAGServices can be used to create Adders with different
+// add implementations.
+type ClusterDAGService interface {
+	ipld.DAGService
+	Finalize(context.Context) (*cid.Cid, error)
+}
+
+// Adder is used to add content to IPFS Cluster using an implementation of
+// ClusterDAGService.
+type Adder struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	dags ClusterDAGService
+
+	params *api.AddParams
+	output chan *api.AddedOutput
+}
+
+// New returns a new Adder with the given ClusterDAGService, add optios and a
+// channel to send updates during the adding process.
+//
+// An Adder may only be used once.
+func New(ctx context.Context, ds ClusterDAGService, p *api.AddParams, out chan *api.AddedOutput) *Adder {
+	ctx2, cancel := context.WithCancel(ctx)
+
+	// Discard all output
+	if out == nil {
+		out = make(chan *api.AddedOutput, 100)
+		go func() {
+			for range out {
+			}
+		}()
+	}
+
+	return &Adder{
+		ctx:    ctx2,
+		cancel: cancel,
+		dags:   ds,
+		params: p,
+		output: out,
+	}
+}
+
+// FromMultipart adds content from a multipart.Reader. The adder will
+// no longer be usable after calling this method.
+func (a *Adder) FromMultipart(r *multipart.Reader) (*cid.Cid, error) {
```

Comment:
Please add a `context.Context` for tracing purposes as this gets called from the restapi.